sui-execution/latest/sui-adapter/src/temporary_store.rs

```diff
@@ -860,100 +837,132 @@ impl<'backing> TemporaryStore<'backing> {
 
     /// Check that this transaction neither creates nor destroys SUI. This should hold for all txes
     /// except the epoch change tx, which mints staking rewards equal to the gas fees burned in the
-    /// previous epoch.  Specifically, this checks two key invariants about storage fees and storage
-    /// rebate:
+    /// previous epoch.  Specifically, this checks two key invariants about storage
+    /// fees and storage rebate:
     ///
     /// 1. all SUI in storage rebate fields of input objects should flow either to the transaction
     ///    storage rebate, or the transaction non-refundable storage rebate
     /// 2. all SUI charged for storage should flow into the storage rebate field of some output
     ///    object
     ///
-    /// If `do_expensive_checks` is true, this will also check a third invariant:
+    /// This function is intended to be called *after* we have charged for
+    /// gas + applied the storage rebate to the gas object, but *before* we
+    /// have updated object versions.
+    pub fn check_sui_conserved(&self, gas_summary: &GasCostSummary) -> Result<(), ExecutionError> {
+        // total amount of SUI in storage rebate of input objects
+        let mut total_input_rebate = 0;
+        // total amount of SUI in storage rebate of output objects
+        let mut total_output_rebate = 0;
+        for (_, input, output) in self.get_modified_objects() {
+            if let Some((_, storage_rebate)) = input {
+                total_input_rebate += storage_rebate;
+            }
+            if let Some(object) = output {
+                total_output_rebate += object.storage_rebate;
+            }
+        }
+
+        if gas_summary.storage_cost == 0 {
```

Comment:
Would it be more certain if we actually introduce a flag to indicate OOG then?

Reply:
yeah thought about it and not sure... I will try though and see how it looks, thanks!